README.md
# MDC
Movies on Direct Connect

## Description

This project is a movie curator for the Direct Connect (like a Netflix™ for DC) platform to easily find movies to download. The program should be able to connect to a DC hub and download movies and also display movie information and cover artwork. The ui should be clean and user friendly.

## Expected features

1. A homepage showing top movies in various genres
2. Search for a movie
3. Information for a selected movie (taken from YIFY api)
4. Related movies based on selected movie (taken from YIFY api)
5. Download resume and restart
6. Download queue
7. ...


## Tools

1. JavaFX
2. Java
3. YTS movie api
4. JDCbot library   (https://github.com/applegrew/jdcbot)
5. JSON parser from json.org


You need JDK (Java Developer Kit), which provides both runtime and the compiler.


If you get any error then probably you don't have JDK 8 or higher installed. Type

    javac -version

to find the version of java compiler you have. If it says that - No such command found or something like that, then you probably don't have JDK installed or it is not properly set up.

**License::** GNU Public License (See enclosed License.txt)



CHANGELOG
The changelog may not document all chnages made. Best way is view diffs.

-- v0.6 2005-09-12 --
* Added UserMenager and User classes for easier work with users
* Added MySQL support
* Added trigger on search request and returning searvh results (both for active and passive search)
* Added KickUser method
* Added WebPageFetcher and GoogleCalculation classes
* Fixed some small bugs

-- v0.7 2008-May-27 --
* Grouped classes into sub-packages
* Renamed UserMenager to UserManager
* Client-Client protocol has been implemented
* Hub-Client protocol enhanced to support EXTENDED protocol features

-- v0.7 2008-May-31 --
* Added BotEventDispatchThread to call particular protected methods of jDCBot in a different thread
* Added DownloadBot (another bot for example)
* Added a package sheriffbot which contains SheriffBot and its related classes
* Fixed Bugs in the code portion that handels download and upload
* Fixed various bugs in the framework

-- v0.7.2 2008-June-2 --
* Added bandwidth limit control classes for download and upload (in org.elite.jdcbot.util)
	# Added DownloadEntityStream
	# Added UploadEntityStream
	# Added RateConstrainer
	# Added ProgressMeter
* Added download/upload cancel facility
* Added DownloadBot2 which demostrates the new download features
* More bug fix, primarily in download related code

-- v1.0 2008-June-18 --
* Added org.elite.jdcbot.shareframework.FilelistConverter
* Added org.elite.jdcbot.shareframework.FileListManager
* Added org.elite.jdcbot.shareframework.FLDir
* Added org.elite.jdcbot.shareframework.FLFile
* Added org.elite.jdcbot.shareframework.FLInterface
* Added org.elite.jdcbot.shareframework.HashException
* Added org.elite.jdcbot.shareframework.HashManager
* Added org.elite.jdcbot.shareframework.HashUser
* Added org.elite.jdcbot.shareframework.SearchResultSet
* Added org.elite.jdcbot.shareframework.SearchSet
* Added org.elite.jdcbot.shareframework.ShareException
* Added org.elite.jdcbot.shareframework.ShareManagerListener
* Added org.elite.jdcbot.shareframework.UploadStreamManager
* Added org.elite.jdcbot.examples.DemoBot - This will be a full fledged client (which will have no 
	UI but will be controllable by another client). Its main purpose is to provide implementational
	guidelines for user of jDCBot
* Added org.elite.jdcbot.framework.BotInterface
* Added org.elite.jdcbot.framework.DownloadCentral
* Added org.elite.jdcbot.framework.GlobalObjects
* Added org.elite.jdcbot.framework.MultiHubsAdapter - To provide easy method to connect to multiple
	hubs
* Added org.elite.jdcbot.framework.UDPInputThread - To listen for UDP packets (which enclose the
	search results)
* Added org.elite.jdcbot.framework.UDPInputThreadTarget
* Renamed org.elite.jdcbot.util.UploadEntityStream to org.elite.jdcbot.util.InputEntityStream
* Renamed org.elite.jdcbot.util.DownloadEntityStream to org.elite.jdcbot.util.OutputEntityStream
	(DownloadEntityStream still exists but it is deprecated and performs no function, it notifies
	the user about this change)
* Added org.elite.jdcbot.util.GlobalFunctions - It has some general purpose functions that is used
	by framework classes and could be used by users of jDCBot too.
* Fixed some major bugs in UploadHandler which made it unable to function at all
* jDCBot now automatically responds to search queries from other users
* Hashing of files, adding and removing files/directories to share has been implemented in ShareManager
* ShareManager can now create XML bzip2 compressed file list
* Can now parse XML bzip2 compressed file lists and convert that into instance of FileListManager,
	which provides features like searching , quering and manipulating file lists. Manipulating
	downloaded file lists of other users is useless though

-- v1.0 2008-June-25 --
* Made many classes thread-safe. In javadoc of all such classes it has been clearly mentioned if
	they are thread safe. Non-thread safe classes have no such comment.
* Fixed bug in UserManager and User where username comparsions used case sensitive comparisions
	whereas DC protocol makes no such requirements, infact hubs seem to use case-insensitive
	comparisions.
* Added EventjDCBot (and two more accompaning classes for convinience and a listener interface). Its
	purpose is to generate events for listeners for events generated by jDCBot (which traditionally
	allows only its sub-class to listen for events).
* Changed the way setDirs should be called. Now it should be called in jDCBot or MultiHubsAdapter
	(if present) intead of directly calling setDirs of ShareManager and DownloadCentral.
* DownloadCentral is now easier and neater to use. Its bugs have been fixed which made it almost
	non-operational.
* Fixed various bugs in ShareManager, User, UserManager, DownloadHandler, DownloadManager,
	UploadHandler, TimeoutInputThread, DemoBot, and more.
* Created few methods meant to be overridden in ShareManager. See its javadoc for there purpose and
	DemoBot for ways to implement them.
* All event methdos in jDCBot (the ones whose names start with 'on' and are overridable) will now
	be run in jDCBot Dispatch Thread. (Previously only a handful ran in that thread.)

-- v1.0 2008-June-29 --
* Fixed more bugs and removed dead-locks. Hopefully they are all resolved.
* Replaced all use of Vector with ArrayList.
* Added more features into DemoBot.

src/com/rskpdev/Main.java
package com.rskpdev;

import java.io.IOException;
import java.io.PrintStream;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import javafx.application.Application;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.stage.Stage;

public class Main
extends Application {

    static UiLogStream log = new UiLogStream();
    static SimpleDateFormat formatter = new SimpleDateFormat("hh:mm:ss");
    static Calendar calendarNow = Calendar.getInstance();
    static BotLogic logic = new BotLogic((PrintStream)log);

    public static void main(String[] args) throws IOException {
        Main.launch((String[])args);
    }

    public void start(Stage primaryStage) throws Exception {
        Parent root = (Parent)FXMLLoader.load((URL)this.getClass().getResource("MDC.fxml"));
        primaryStage.setTitle("MDC");
        primaryStage.setScene(new Scene(root, 1200.0, 500.0));
        primaryStage.show();
        primaryStage.setOnCloseRequest(event ->DisconnectDC()); //disconnect form DC when closing window
        this.ConnectToDC(); //connect to DC on start (should change to connect on button click)
    }

    public void ConnectToDC() {

        //edit the below parameters for your hub (change so that this is user editable in a setting window)
        logic.configBot("MovieBot" //Bot name
                , "192.168.210.116", //Bot IP
                9000, //Listen port
                10000, //listen port UDP
                "", //password
                "rskpbot", //description
                "ADSL1", //connection type
                "@", //email
                "0", //share size
                2, //upload slots
                4, //download slots
                true, //active mode
                "C:/Users/Rahul/Desktop/tmp", //download folder (must exist, or crashes)
                "C:/Users/Rahul/Desktop/tmp/Incomplete");//incomplete downloads folder (must exist, or crashes)


        //hub IP and port(also change so that this is user editable in a setting window)
        if (logic.connect("192.168.1.4", 411)) {
            log.println(formatter.format(calendarNow.getTime()) + ": Is connected now");
        }
    }

    private void DisconnectDC() {
        logic.disconnect();
        log.println(formatter.format(calendarNow.getTime()) + ": Is disconnected now");
    }

    public static void SearchMovie(String name) {
        logic.search(name);
        log.println(formatter.format(calendarNow.getTime()) + ": Search finished");
    }

    public static void DownloadMovie(String magnet) {
        logic.downloadFile(magnet); //Downloading needs a magnet URL
        log.println(formatter.format(calendarNow.getTime()) + ": Starting download process");
    }
}

src/com/rskpdev/Parser.java
package com.rskpdev;

import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Created by Rahul on 07/14/16.
 * this class is used to hold code for  parsing JSON strings
 * parser is taken from json.org (http://mvnrepository.com/artifact/org.json/json)
 * import the json library for it to work
 */
public class Parser{

    /*
    gets an array of urls of the movies covers given a list of movie in JSON
     */
    public String[] coverParse(String input){
        JSONObject obj = new JSONObject(input);
        JSONArray arr = obj.getJSONObject("data").getJSONArray("movies");
        String[] output = new String[arr.length()];

        for (int i = 0; i < arr.length(); i++)
        {
            output[i] = arr.getJSONObject(i).getString("large_cover_image");
        }

        return output;
    }

    //gets the title of a single movie
    public String titleParse(String input){
        JSONObject obj = new JSONObject(input);
        JSONArray arr = obj.getJSONObject("data").getJSONArray("movies");
        String output = arr.getJSONObject(0).getString("title");
        return output;
    }

    //gets the year of a single movie
    public int yearParse(String input){
        JSONObject obj = new JSONObject(input);
        JSONArray arr = obj.getJSONObject("data").getJSONArray("movies");
        int output = arr.getJSONObject(0).getInt("year");
        return output;
    }

    //gets the rating (IMDB) of a single movie
    public double ratingParse(String input){
        JSONObject obj = new JSONObject(input);
        JSONArray arr = obj.getJSONObject("data").getJSONArray("movies");
        double output = arr.getJSONObject(0).getDouble("rating");           //rating is given in double
        return output;
    }

    //gets the summary of a single movie
    public String summaryParse(String input){
        JSONObject obj = new JSONObject(input);
        JSONArray arr = obj.getJSONObject("data").getJSONArray("movies");
        String output = arr.getJSONObject(0).getString("summary");
        return output;
    }

}


src/com/rskpdev/Controller.java
package com.rskpdev;

import com.rskpdev.Main;
import com.rskpdev.Parser;
import com.rskpdev.YtsFetcher;
import java.io.IOException;
import java.net.URL;
import java.util.ResourceBundle;
import javafx.event.Event;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;

public class Controller
implements Initializable {
    @FXML
    TextField search_box;
    @FXML
    Label title;
    @FXML
    Label year;
    @FXML
    Label summary;
    @FXML
    Label rating;
    @FXML
    Label available;

    public void initialize(URL location, ResourceBundle resources) {
    }

    //search button is clicked
    public void searchClicked() throws IOException {
        //find movie on YIFY
        //if field is left empty it give a random movie ( fix this)
        String movie = YtsFetcher.searchMovie((String)this.search_box.getText());

        Parser parse = new Parser();
        //set all fields
        this.summary.setText(parse.summaryParse(movie));
        this.title.setText(parse.titleParse(movie));
        this.year.setText(String.valueOf(parse.yearParse(movie)));
        this.rating.setText(String.valueOf(parse.ratingParse(movie)) + "/10 IMDB");
        //search for the movie on DC
        Main.SearchMovie(this.title.getText());
    }

    public void download(Event event) {
    }
}

src/com/rskpdev/YtsFetcher.java
package com.rskpdev;

import org.json.JSONObject;
import java.io.BufferedReader;
import org.json.*;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;
import java.nio.charset.Charset;

/**
 * Created by Rahul on 07/14/16.
 * Fetches JSON strings from YIFY  api
 * needs a working internet connection or causes a crash(must be fixed)
 */
public class YtsFetcher {

    private static String top_movies;

    //fetches top movies from yify
    public static void fetchTop(int page, int limit) throws IOException {
        top_movies = fetchResult("https://yts.ag/api/v2/list_movies.json?limit="+limit+"&sort_by=download_count&page="+page);
    }

    public static String getTop(){
        return top_movies;
    }

    public static String searchMovie(String name) throws IOException {

        //cannot handle movies with spaces in title (fix this)
        return fetchResult("https://yts.ag/api/v2/list_movies.json?query_term=" + name);
    }



    //fetches JSON String from query url
    private static String fetchResult(String query) throws IOException {
        URLConnection connection = new URL(query).openConnection();
        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11");
        connection.connect();

        BufferedReader r  = new BufferedReader(new InputStreamReader(connection.getInputStream(), Charset.forName("UTF-8")));

        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = r.readLine()) != null) {
            sb.append(line);
        }
        return sb.toString();
    }

}


src/org/elite/jdcbot/examples/BotMain.java
package org.elite.jdcbot.examples;

import java.awt.*;
import java.awt.event.*;
import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintStream;
import java.text.*;
import java.util.*;
import javax.swing.*;
import javax.swing.border.*;

import org.elite.jdcbot.framework.User;

public class BotMain extends JFrame implements ActionListener {
	
	private static final long serialVersionUID = 1L;

	// GUI
	JMenuBar mnuBar = new JMenuBar();
	JMenu mnuFile = new JMenu("File");
	JMenuItem itemExit = new JMenuItem("Exit");
	JMenu mnuHelp = new JMenu("Help");
	JMenuItem itemAbout = new JMenuItem("About");
	
	JPanel pnlServer = new JPanel();
	JLabel lblServer = new JLabel("Server : Port");
	JTextField fieldServerDC = new JTextField("192.168.179.128", 11);
	JTextField fieldServerPortDC = new JTextField("411", 3);
	JButton btnConnect = new JButton("Connect");
	JButton btnDisconnect = new JButton("Disconnect");
	
	JPanel pnlDirectories = new JPanel();
	JPanel pnlDir1 = new JPanel();
	JLabel lblDir1 = new JLabel("File path: ");
	JTextField fieldDir1 = new JTextField("C:/tmp",20);
	JPanel pnlDir2 = new JPanel();
	JLabel lblDir2 = new JLabel("File path: ");
	JTextField fieldDir2 = new JTextField("C:/tmp/Incomplete",20); 

	JTextArea txtLog = new JTextArea();
	JScrollPane scrollPane = new JScrollPane(txtLog);

	JPanel pnlUsers = new JPanel();
	JTextField fieldUser = new JTextField("njmartinez");
	JButton btnGetUserFiles = new JButton("Get user files");
	JTextField fieldSearch = new JTextField();
	JButton btnSearch = new JButton("Search");
	JTextField fieldFile = new JTextField();
	JButton btnDownload = new JButton("Download");
	JFileChooser chooser = new JFileChooser();
	JButton btnAddDir = new JButton("Add directory");
	JButton btnExcludeDir = new JButton("Exclude directory");
	JButton btnRemoveDir = new JButton("Remove directory");
	JButton btnProcessList = new JButton("Process list");

	JPanel pnlConfig = new JPanel();
	JLabel lblBotName = new JLabel("Bot name");
	JTextField fieldBotName = new JTextField("miBot");
	JLabel lblBotIp = new JLabel("Bot Ip");
	JTextField fieldBotIp = new JTextField("192.168.210.116");
	JLabel lblListenPort = new JLabel("Port");
	JTextField fieldListenPort = new JTextField("9000");
	JLabel lblListenPortUDP = new JLabel("UDP port");
	JTextField fieldListenPortUDP = new JTextField("10000");
	JLabel lblPassword = new JLabel("Password");
	JTextField fieldPassword = new JTextField("");
	JLabel lblDescription = new JLabel("Description");
	JTextField fieldDescription = new JTextField("Bot de pruebas");
	JLabel lblConnectionType = new JLabel("Connection Type");
	JTextField fieldConnectionType = new JTextField("ADSL");
	JLabel lblEmail = new JLabel("Email");
	JTextField fieldEmail = new JTextField("@");
	JLabel lblShareSize = new JLabel("Share size");
	JTextField fieldShareSize = new JTextField("0");
	JLabel lblUploadSlots = new JLabel("Upload slots");
	JTextField fieldUploadSlots = new JTextField("2");
	JLabel lblDownloadSlots = new JLabel("Download slots");
	JTextField fieldDownloadSlots = new JTextField("4");
	JLabel lblisPassive = new JLabel("Set mode");
	String [] states = {"Active", "Passive"};
	JComboBox cmbStates = new JComboBox(states);
	
	JTextField fieldInformation = new JTextField();
	
	// Common
	UiLogStream log = new UiLogStream();
	
	// Logic
	private BotLogic logic = new BotLogic(log);

	public BotMain() {
		
		setTitle("ClienteBot Direct Connect");
		setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
		setResizable(true);
		setJMenuBar(mnuBar);
		
		btnDisconnect.setEnabled(false);
		txtLog.setEditable(false);
		scrollPane.setPreferredSize(new Dimension(250, 250));
		fieldInformation.setEditable(false);
		cmbStates.setSelectedIndex(1);
		
		TitledBorder pnlBorderConfig, pnlBorderServ, pnlBorderDir1, pnlBorderDir2;
		pnlBorderConfig = BorderFactory.createTitledBorder("Bot configuration");
		pnlBorderServ = BorderFactory.createTitledBorder("Quick connect");
		pnlBorderDir1 = BorderFactory.createTitledBorder("Config directory");
		pnlBorderDir2 = BorderFactory.createTitledBorder("Temporal download directory");
		pnlConfig.setBorder(pnlBorderConfig);
		pnlServer.setBorder(pnlBorderServ);
		pnlDir1.setBorder(pnlBorderDir1);
		pnlDir2.setBorder(pnlBorderDir2);

		JPanel pnlGeneral = (JPanel) this.getContentPane();

		mnuBar.add(mnuFile);
		mnuFile.add(itemExit);
		mnuBar.add(mnuHelp);
		mnuHelp.add(itemAbout);
		
		pnlServer.add(lblServer);
		pnlServer.add(fieldServerDC);
		pnlServer.add(fieldServerPortDC);
		pnlServer.add(btnConnect);
		pnlServer.add(btnDisconnect);
		
		pnlDir1.add(lblDir1);
		pnlDir1.add(fieldDir1);
		pnlDir2.add(lblDir2);
		pnlDir2.add(fieldDir2);
		pnlDirectories.add(pnlDir1);
		pnlDirectories.add(pnlDir2);
		
		JTabbedPane tabbedPaneS = new JTabbedPane();
		tabbedPaneS.addTab("Server", pnlServer);
		tabbedPaneS.addTab("Directories", pnlDirectories);
		
		pnlUsers.setLayout(new GridLayout(0, 2));
		pnlUsers.add(fieldUser);
		pnlUsers.add(btnGetUserFiles);
		pnlUsers.add(fieldSearch);
		pnlUsers.add(btnSearch);
		pnlUsers.add(fieldFile);
		pnlUsers.add(btnDownload);
		pnlUsers.add(btnAddDir);
		pnlUsers.add(btnExcludeDir);
		pnlUsers.add(btnRemoveDir);
		pnlUsers.add(btnProcessList);
		
		pnlConfig.setLayout(new GridLayout(0, 2));
		pnlConfig.add(lblBotName);
		pnlConfig.add(fieldBotName);
		pnlConfig.add(lblBotIp);
		pnlConfig.add(fieldBotIp);
		pnlConfig.add(lblListenPort);
		pnlConfig.add(fieldListenPort);
		pnlConfig.add(lblListenPortUDP);
		pnlConfig.add(fieldListenPortUDP);
		pnlConfig.add(lblPassword);
		pnlConfig.add(fieldPassword);
		pnlConfig.add(lblDescription);
		pnlConfig.add(fieldDescription);
		pnlConfig.add(lblConnectionType);
		pnlConfig.add(fieldConnectionType);
		pnlConfig.add(lblEmail);
		pnlConfig.add(fieldEmail);
		pnlConfig.add(lblShareSize);
		pnlConfig.add(fieldShareSize);
		pnlConfig.add(lblUploadSlots);
		pnlConfig.add(fieldUploadSlots);
		pnlConfig.add(lblDownloadSlots);
		pnlConfig.add(fieldDownloadSlots);
		pnlConfig.add(lblisPassive);
		pnlConfig.add(cmbStates);		
		
		JTabbedPane tabbedPane = new JTabbedPane();
		tabbedPane.addTab("Actions", pnlUsers);
		tabbedPane.addTab("Log", scrollPane);
		
		pnlGeneral.add("North", tabbedPaneS);
		pnlGeneral.add("Center", tabbedPane);
		pnlGeneral.add("West", pnlConfig);
		pnlGeneral.add("South", fieldInformation);
		setContentPane(pnlGeneral);
		pack();

		itemExit.addActionListener(this);
		btnConnect.addActionListener(this);
		btnDisconnect.addActionListener(this);
		btnGetUserFiles.addActionListener(this);
		btnSearch.addActionListener(this);
		btnDownload.addActionListener(this);
		btnAddDir.addActionListener(this);
		btnExcludeDir.addActionListener(this);
		btnRemoveDir.addActionListener(this);
		btnProcessList.addActionListener(this);
		
		chooser.setCurrentDirectory(new java.io.File("."));
		chooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
		chooser.setAcceptAllFileFilterUsed(false);
	}

	/** 
	 * Handles the buttons' actions.
	 */
	public void actionPerformed(ActionEvent ae) {
		SimpleDateFormat formatter = new SimpleDateFormat("hh:mm:ss");
		Calendar calendarNow = Calendar.getInstance();
		
		if(ae.getSource() == itemExit){
			if(btnDisconnect.isEnabled()){
				logic.disconnect();
			}
			System.exit(0);
		}
		else if (ae.getSource() == btnConnect) {
			boolean isActive = false;
			if(cmbStates.getSelectedItem() == "Active"){
				isActive = true;
			}
			else if (cmbStates.getSelectedItem() == "Passive"){
				isActive = false;
			}
			logic.configBot(fieldBotName.getText(), fieldBotIp.getText(),
					Integer.parseInt(fieldListenPort.getText()),
					Integer.parseInt(fieldListenPortUDP.getText()),
					fieldPassword.getText(), fieldDescription.getText(),
					fieldConnectionType.getText() + User.NORMAL_FLAG,
					fieldEmail.getText(), fieldShareSize.getText(),
					Integer.parseInt(fieldUploadSlots.getText()),
					Integer.parseInt(fieldDownloadSlots.getText()),
					isActive, fieldDir1.getText(), fieldDir2.getText());
			if (logic.connect(fieldServerDC.getText(), Integer.parseInt(fieldServerPortDC.getText()))) {
				btnConnect.setEnabled(false);
				btnDisconnect.setEnabled(true);
				log.println(formatter.format(calendarNow.getTime()) + ": Is connected now");
				fieldInformation.setText("Connected");
			}
		} else if (ae.getSource() == btnDisconnect) {
			logic.disconnect();
			btnConnect.setEnabled(true);
			btnDisconnect.setEnabled(false);
			log.println(formatter.format(calendarNow.getTime()) + ": Is disconnected now");
			fieldInformation.setText("Disconnected");
		} else if (ae.getSource() == btnGetUserFiles) {
			logic.readFiles(fieldUser.getText());
			log.println(formatter.format(calendarNow.getTime()) + ": Get user files finished");
			fieldInformation.setText("Get user files finished");
		} else if (ae.getSource() == btnSearch) {
			String searchText = fieldSearch.getText();
			if (searchText.isEmpty()) {
				fieldInformation.setText("Enter a word to start searching");
			} else {
				logic.search(searchText);
				log.println(formatter.format(calendarNow.getTime()) + ": Search finished");
				fieldInformation.setText("Search finished");
			}
		} else if (ae.getSource()==btnDownload){
			logic.downloadFile(fieldFile.getText());
			log.println(formatter.format(calendarNow.getTime()) + ": Starting download process");
			fieldInformation.setText("Processing a download");
		}else if(ae.getSource() == btnAddDir){		
			int selection = chooser.showOpenDialog(this);
			if (selection == JFileChooser.APPROVE_OPTION) {
				String directory = chooser.getSelectedFile().toString();
				logic.addDirectory(directory);
				JOptionPane.showMessageDialog(getRootPane(), "This folder has been added to the list:\n" + directory);
				log.println(formatter.format(calendarNow.getTime())+": This folder has been added to the list: "+ directory);
				fieldInformation.setText("Folder adicionado");
			}	  
		} else if(ae.getSource() == btnExcludeDir){
			int selection = chooser.showOpenDialog(this);
			if (selection == JFileChooser.APPROVE_OPTION) {
				String directory = chooser.getSelectedFile().toString();
				logic.excludeDirectory(directory);
				JOptionPane.showMessageDialog(getRootPane(), "This folder has beend excluded from the list:\n" + directory);
				log.println(formatter.format(calendarNow.getTime())+": This folder has beend excluded from the list: "+ directory);
				fieldInformation.setText("Folder excluido");
			}
		} else if(ae.getSource() == btnRemoveDir){
			int selection = chooser.showOpenDialog(this);
			if (selection == JFileChooser.APPROVE_OPTION) {
				String directory = chooser.getSelectedFile().toString();
				logic.removeDirectory(directory);
				JOptionPane.showMessageDialog(getRootPane(), "This folder has been removed from the list:\n" + directory);
				log.println(formatter.format(calendarNow.getTime())+": This folder has been removed from the list: "+ directory);
				fieldInformation.setText("Folder retirado");
			}
		} else if(ae.getSource() == btnProcessList){
			logic.processList();
			JOptionPane.showMessageDialog(getRootPane(), "The list has been processed");
			log.println(formatter.format(calendarNow.getTime())+": The list has been updated");
			fieldInformation.setText("List updated");
		}
	}
	
	private class UiLogStream extends PrintStream {
		public UiLogStream() {
			super(new OutputStream() {
				StringBuffer buffer = new StringBuffer();

				@Override
				public synchronized void write(int ch) throws IOException {
					buffer.append((char) ch);
				}
				
				@Override
				public synchronized void flush() {
					txtLog.append(buffer.toString());
					buffer = new StringBuffer();
				}
			}, true);
		}
	}

	// main
	public static void main(String[] args) {
		try {
			UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
		} catch (Exception e) {
			e.printStackTrace();
		}
		BotMain clienteBot = new BotMain();
		clienteBot.setVisible(true);
	}	
}
